### PR TITLE
#1741 Experimental visible text condition

### PR DESCRIPTION
**Pull request by aakachurin**
## Proposed changes
For #1741
I don't know how it works with react or another framework
## Checklist
- [x] Checkstyle and unit tests are passed locally with my changes by running `gradlew check chrome_headless firefox_headless` command
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: src/main/java/com/codeborne/selenide/conditions/VisibleText.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
+package com.codeborne.selenide.conditions;
+
+import com.codeborne.selenide.Driver;
+import com.codeborne.selenide.impl.JavaScript;
+import org.openqa.selenium.WebElement;
+
+import javax.annotation.Nullable;
+
+public class VisibleText extends CaseInsensitiveTextCondition {
```
Comment (aakachurin) [inline src/main/java/com/codeborne/selenide/conditions/VisibleText.java:9]: I think it's better to extend it from TextCondition
What do you think

**Comment by aakachurin** [comment]
@BorisOsipov @asolntsev could you review it? 
And there are problems with checks, but i don't see any errors

**Comment by BorisOsipov** [comment]
>but i don't see any errors

```
2023-01-23T22:57:34.4283664Z ArchitecturalChecks STANDARD_ERROR
2023-01-23T22:57:34.4284289Z     22:57:14:048 [Test worker] INFO PluginLoader - Detected Java version 17.0.5
2023-01-23T22:57:34.4284492Z 
2023-01-23T22:57:34.4284633Z ArchitecturalChecks > noJunitAsserts() FAILED
2023-01-23T22:57:34.4285420Z     java.lang.AssertionError: Architecture Violation [Priority: MEDIUM] - Rule 'no classes should access classes that are assignable to org.junit.jupiter.api.Assertions' was violated (1 times):
2023-01-23T22:57:34.4286344Z     Method <integration.VisibleTextTest.shouldNotMatchExpectedVisibleTextWithDifferentCase2()> calls method <org.junit.jupiter.api.Assertions.assertThrows(java.lang.Class, org.junit.jupiter.api.function.Executable)> in (VisibleTextTest.java:30)
2023-01-23T22:57:34.4287180Z         at com.tngtech.archunit.lang.ArchRule$Assertions.assertNoViolation(ArchRule.java:94)
2023-01-23T22:57:34.4287625Z         at com.tngtech.archunit.lang.ArchRule$Assertions.check(ArchRule.java:86)
2023-01-23T22:57:34.4288066Z         at com.tngtech.archunit.lang.ArchRule$Factory$SimpleArchRule.check(ArchRule.java:165)
2023-01-23T22:57:34.4288587Z         at com.tngtech.archunit.lang.syntax.ObjectsShouldInternal.check(ObjectsShouldInternal.java:81)
2023-01-23T22:57:34.4289143Z         at com.codeborne.selenide.ArchitecturalChecks.noJunitAsserts(ArchitecturalChecks.java:17)
```

firefox_headless:

```
2023-01-23T23:01:56.8314638Z VisibleTextTest > shouldMatchExpectedVisibleTextWithDifferentCase() FAILED
2023-01-23T23:01:56.8315759Z     Element should Case-insensitive visible text "987 65423 2aa21" {[name='field_value']}
2023-01-23T23:01:56.8316312Z     Element: '<span name="field_value">987 65423 2Aa21 1 10asdasda10 100.876543321 AUTOOQODS</span>'
2023-01-23T23:01:56.8316660Z     Actual value: text="987 65423 2Aa21 1 10asdasda10 100.876543321"
2023-01-23T23:01:56.8317372Z     Screenshot: file:/home/runner/work/selenide/selenide/modules/core/build/reports/tests/firefox_headless/1674514916582.135.png
2023-01-23T23:01:56.8317872Z     Page source: file:/home/runner/work/selenide/selenide/modules/core/build/reports/tests/firefox_headless/1674514916582.135.html
2023-01-23T23:01:56.8318206Z     Timeout: 4 s.
2023-01-23T23:01:56.8318600Z         at app//com.codeborne.selenide.impl.WebElementSource.handleError(WebElementSource.java:146)
2023-01-23T23:01:56.8319230Z         at app//com.codeborne.selenide.impl.WebElementSource.checkConditionAndReturnElement(WebElementSource.java:124)
2023-01-23T23:01:56.8319844Z         at app//com.codeborne.selenide.impl.WebElementSource.checkCondition(WebElementSource.java:101)
2023-01-23T23:01:56.8320334Z         at app//com.codeborne.selenide.commands.Should.execute(Should.java:30)
2023-01-23T23:01:56.8320767Z         at app//com.codeborne.selenide.commands.Should.execute(Should.java:14)
2023-01-23T23:01:56.8321214Z         at app//com.codeborne.selenide.commands.Commands.execute(Commands.java:159)
2023-01-23T23:01:56.8321760Z         at app//com.codeborne.selenide.impl.SelenideElementProxy.dispatchAndRetry(SelenideElementProxy.java:131)
2023-01-23T23:01:56.8322341Z         at app//com.codeborne.selenide.impl.SelenideElementProxy.invoke(SelenideElementProxy.java:83)
2023-01-23T23:01:56.8322785Z         at app/jdk.proxy3/jdk.proxy3.$Proxy31.should(Unknown Source)
2023-01-23T23:01:56.8323280Z         at app//integration.VisibleTextTest.lambda$shouldMatchExpectedVisibleTextWithDifferentCase$1(VisibleTextTest.java:25)
2023-01-23T23:01:56.8323765Z         at app//integration.ITest.withLongTimeout(ITest.java:34)
2023-01-23T23:01:56.8324289Z         at app//integration.VisibleTextTest.shouldMatchExpectedVisibleTextWithDifferentCase(VisibleTextTest.java:25)
```

```
2023-01-23T23:02:01.0321121Z VisibleTextTest > shouldMatchExpectedVisibleTextWithSameCase2() FAILED
2023-01-23T23:02:01.0321626Z     Element should Case sensitive visible text "987 65423 2Aa21" {[name='field_value']}
2023-01-23T23:02:01.0322101Z     Element: '<span name="field_value">987 65423 2Aa21 1 10asdasda10 100.876543321 AUTOOQODS</span>'
2023-01-23T23:02:01.0322449Z     Actual value: text="987 65423 2Aa21 1 10asdasda10 100.876543321"
2023-01-23T23:02:01.0322856Z     Screenshot: file:/home/runner/work/selenide/selenide/modules/core/build/reports/tests/firefox_headless/1674514920899.136.png
2023-01-23T23:02:01.0323345Z     Page source: file:/home/runner/work/selenide/selenide/modules/core/build/reports/tests/firefox_headless/1674514920899.136.html
2023-01-23T23:02:01.0323678Z     Timeout: 4 s.
2023-01-23T23:02:01.0324089Z         at app//com.codeborne.selenide.impl.WebElementSource.handleError(WebElementSource.java:146)
2023-01-23T23:02:01.0324710Z         at app//com.codeborne.selenide.impl.WebElementSource.checkConditionAndReturnElement(WebElementSource.java:124)
2023-01-23T23:02:01.0325321Z         at app//com.codeborne.selenide.impl.WebElementSource.checkCondition(WebElementSource.java:101)
2023-01-23T23:02:01.0325812Z         at app//com.codeborne.selenide.commands.Should.execute(Should.java:30)
2023-01-23T23:02:01.0326240Z         at app//com.codeborne.selenide.commands.Should.execute(Should.java:14)
2023-01-23T23:02:01.0326685Z         at app//com.codeborne.selenide.commands.Commands.execute(Commands.java:159)
2023-01-23T23:02:01.0327230Z         at app//com.codeborne.selenide.impl.SelenideElementProxy.dispatchAndRetry(SelenideElementProxy.java:131)
2023-01-23T23:02:01.0327803Z         at app//com.codeborne.selenide.impl.SelenideElementProxy.invoke(SelenideElementProxy.java:83)
2023-01-23T23:02:01.0328258Z         at app/jdk.proxy3/jdk.proxy3.$Proxy31.should(Unknown Source)
2023-01-23T23:02:01.0328742Z         at app//integration.VisibleTextTest.lambda$shouldMatchExpectedVisibleTextWithSameCase2$3(VisibleTextTest.java:38)
2023-01-23T23:02:01.0329218Z         at app//integration.ITest.withLongTimeout(ITest.java:34)
2023-01-23T23:02:01.0329732Z         at app//integration.VisibleTextTest.shouldMatchExpectedVisibleTextWithSameCase2(VisibleTextTest.java:38)
```

```
2023-01-23T23:02:05.3347482Z VisibleTextTest > shouldMatchExpectedVisibleTextWithSameCase() FAILED
2023-01-23T23:02:05.3348187Z     Element should Case-insensitive visible text "987 65423 2Aa21" {[name='field_value']}
2023-01-23T23:02:05.3348761Z     Element: '<span name="field_value">987 65423 2Aa21 1 10asdasda10 100.876543321 AUTOOQODS</span>'
2023-01-23T23:02:05.3349185Z     Actual value: text="987 65423 2Aa21 1 10asdasda10 100.876543321"
2023-01-23T23:02:05.3349668Z     Screenshot: file:/home/runner/work/selenide/selenide/modules/core/build/reports/tests/firefox_headless/1674514925205.137.png
2023-01-23T23:02:05.3350240Z     Page source: file:/home/runner/work/selenide/selenide/modules/core/build/reports/tests/firefox_headless/1674514925205.137.html
2023-01-23T23:02:05.3350650Z     Timeout: 4 s.
2023-01-23T23:02:05.3351143Z         at app//com.codeborne.selenide.impl.WebElementSource.handleError(WebElementSource.java:146)
2023-01-23T23:02:05.3351883Z         at app//com.codeborne.selenide.impl.WebElementSource.checkConditionAndReturnElement(WebElementSource.java:124)
2023-01-23T23:02:05.3352614Z         at app//com.codeborne.selenide.impl.WebElementSource.checkCondition(WebElementSource.java:101)
2023-01-23T23:02:05.3353200Z         at app//com.codeborne.selenide.commands.Should.execute(Should.java:30)
2023-01-23T23:02:05.3353713Z         at app//com.codeborne.selenide.commands.Should.execute(Should.java:14)
2023-01-23T23:02:05.3354249Z         at app//com.codeborne.selenide.commands.Commands.execute(Commands.java:159)
2023-01-23T23:02:05.3354912Z         at app//com.codeborne.selenide.impl.SelenideElementProxy.dispatchAndRetry(SelenideElementProxy.java:131)
2023-01-23T23:02:05.3355590Z         at app//com.codeborne.selenide.impl.SelenideElementProxy.invoke(SelenideElementProxy.java:83)
2023-01-23T23:02:05.3356113Z         at app/jdk.proxy3/jdk.proxy3.$Proxy31.should(Unknown Source)
2023-01-23T23:02:05.3358018Z         at app//integration.VisibleTextTest.lambda$shouldMatchExpectedVisibleTextWithSameCase$0(VisibleTextTest.java:20)
2023-01-23T23:02:05.3358593Z         at app//integration.ITest.withLongTimeout(ITest.java:34)
2023-01-23T23:02:05.3359226Z         at app//integration.VisibleTextTest.shouldMatchExpectedVisibleTextWithSameCase(VisibleTextTest.java:20)
```

**Comment by aakachurin** [comment]
I did the research and realized that the resulting text may be different depending on platforms and browsers, but the functionality still works. It helped to specify the font explicitly

@BorisOsipov Could you review it again?

**Comment by BorisOsipov** [comment]
To be honest, I am skeptical about this feature. [According to the WebDriver specification,](https://www.w3.org/TR/webdriver/#get-element-text) the getText() method should return:

```Let rendered text be the result of performing implementation-specific steps whose result is exactly the same as the result of a Function.[[Call]](null, element) with bot.dom.getVisibleText as the this value.```


Therefore, it must return visible text. As you can see, [bot.dom.getVisibleText ](https://github.com/SeleniumHQ/selenium/blob/a6b161a159c3d581b130f03a2e6e35f577f38dec/javascript/atoms/dom.js#L1007)is complex and covers different corner cases.

How can I see that the current implementation doesn't cover simple cases, e.g.:
Let's say we have HTML from a unit test, and I modified it a bit:
```
<div id="test">
  <div style="display:block;height:19px;max-width:132.5px;overflow:hidden;text-overflow:ellipsis;font-family: Arial, serif">
    <div style="display:block;height:19px;max-width:172.5px;overflow:hidden;text-overflow:ellipsis;font-family: Arial, serif">
      <span name="field_value">987 65423 2Aa21 1 10asdasda10 100.876543321 AUTOOQODS</span>
    </div>
  </div>
</div>
```
1 - `("[name='field_value']").should(Condition.visibleText("987 65423 2Aa21"))` - works
2 - `$("#test").should(Condition.visibleText("987 65423 2Aa21"))` - fails.

But I expect 1 and 2 to work identically.

I believe there are more of such cases that are expected to work but do not. I think it is not a good idea to re-implement bot.dom.getVisibleText to cover all of them.

**Comment by aakachurin** [comment]
> To be honest, I am skeptical about this feature. [According to the WebDriver specification,](https://www.w3.org/TR/webdriver/#get-element-text) the getText() method should return:
> 
> `Let rendered text be the result of performing implementation-specific steps whose result is exactly the same as the result of a Function.[[Call]](null, element) with bot.dom.getVisibleText as the this value.`
> 
> Therefore, it must return visible text. As you can see, [bot.dom.getVisibleText ](https://github.com/SeleniumHQ/selenium/blob/a6b161a159c3d581b130f03a2e6e35f577f38dec/javascript/atoms/dom.js#L1007)is complex and covers different corner cases.
> 
> How can I see that the current implementation doesn't cover simple cases, e.g.: Let's say we have HTML from a unit test, and I modified it a bit:
> 
> ```
> <div id="test">
>   <div style="display:block;height:19px;max-width:132.5px;overflow:hidden;text-overflow:ellipsis;font-family: Arial, serif">
>     <div style="display:block;height:19px;max-width:172.5px;overflow:hidden;text-overflow:ellipsis;font-family: Arial, serif">
>       <span name="field_value">987 65423 2Aa21 1 10asdasda10 100.876543321 AUTOOQODS</span>
>     </div>
>   </div>
> </div>
> ```
> 
> 1 - `("[name='field_value']").should(Condition.visibleText("987 65423 2Aa21"))` - works 2 - `$("#test").should(Condition.visibleText("987 65423 2Aa21"))` - fails.
> 
> But I expect 1 and 2 to work identically.
> 
> I believe there are more of such cases that are expected to work but do not. I think it is not a good idea to re-implement bot.dom.getVisibleText to cover all of them.

I agree, we can just rename the method to getVisibleNodeText or something like else.
Maybe it will be you useful for someone.
Or we can just close this pr :D

**Comment by asolntsev** [comment]
@aakachurin Honestly, I am not sure about anything. Does it work in all cases? Do we have the problem to solve at all?..

Well, probably it's simpler to close the PR for now.
Thank you for the contribution, I value it. :) 
Maybe we can come back to this topic later.
